feat: Support disjunctive authorization checks (Search layer)

ce-dmelnych opened the pull request:
> [!NOTE]  
> This PR is relatively large and includes not only the core implementation of composite (ANY) authorization checks, but also some related refactorings and improvements.
The commits are structured by concern. Reviewing it commit‑by‑commit is recommended to more easily follow the core changes.

## Description

This pull request introduces support for composite (disjunctive) authorization checks in the Search layer, and wires this capability specifically for user task queries.

The current implementation allows enforcing access control using a **single** `Authorization` per request. This is not sufficient in scenarios where access should be granted if **any** of several authorization branches applies. A key example is user task querying, where as part of https://github.com/camunda/camunda/issues/39992 epic access should be allowed when the caller has:
- process‑definition–level permissions (`PROCESS_DEFINITION[READ_USER_TASK]`), or
- task‑level permissions (`USER_TASK[READ]`).

### Key Changes

**AuthorizationCondition abstraction**
- Introduced `AuthorizationCondition` interface with `SingleAuthorizationCondition` and `AnyOfAuthorizationCondition` implementations
- Updated `SecurityContext` to use `AuthorizationCondition` instead of direct `Authorization`, maintaining backward compatibility through builder methods
- Enhanced `AuthorizationCheck` to hold `AuthorizationCondition` for composite checks
- Enhanced `ResourceAccessChecks` to hold `AuthorizationCondition` for composite checks
- Added composite-aware `resourceId` resolution in `ResourceAccessChecks` via `getAuthorizedResourceIdsByType()`
- Decoupled `AuthorizationChecker` from `SecurityContext` for cleaner separation of concerns

**Exception Handling improvements**
- Enhanced `ResourceAccessDeniedException` to support multiple missing authorizations with improved error messages
- Fixed `TenantAccessDeniedException` to include `FORBIDDEN` reason (instead of default `UNKNOWN`)

**Shared search layer support**
- Added `AnyOfAuthorizationCondition` support for both `doGet` and `doSearch` operations in `AbstractResourceAccessController`

**ES/OS layer support**
- Extended `IndexFilterTransformer` to support composite authorization conditions
- Updated `UserTaskFilterTransformer` to support `USER_TASK` resource type checks

**RDBMS layer support**
- Introduced dedicated authorization fields in `UserTaskDbQuery` (`authorizedProcessDefinitionIds`, `authorizedUserTaskKeys`)
- Updated `UserTaskMapper` to apply per-type authorization filters

### How to use it

#### Example: user task search with composite (ANY) authorization

Currently, a user task search endpoint enforces a single `PROCESS_DEFINITION[READ_USER_TASK]` permission check:

```java
@Override
public SearchQueryResult<UserTaskEntity> search(final UserTaskQuery query) {
  return search(
      query,
      securityContextProvider.provideSecurityContext(
          authentication, Authorization.<UserTaskEntity>of(a -> a.processDefinition().readUserTask())));
}
```

With composite checks, we can enforce to grant access when either of these 2 permissions caller has:
- `PROCESS_DEFINITION[READ_USER_TASK]`, or
- `USER_TASK[READ]`

by changing the code to:
```java
@Override
public SearchQueryResult<UserTaskEntity> search(final UserTaskQuery query) {
  return search(
      query,
      securityContextProvider.provideSecurityContext(
          authentication,
          AuthorizationConditions.anyOf(
              Authorization.<UserTaskEntity>of(a -> a.processDefinition().readUserTask()),
              Authorization.<UserTaskEntity>of(a -> a.userTask().read()))));
}
```

### Adoption guidelines

This section explains how to adopt composite authorization checks with different `AuthorizationResourceType`s for additional resources beyond user tasks.

#### 1. ES/OS implementation
To support composite checks spanning multiple authorizations with different `AuthorizationResourceType`s for a given index(resource), you need to adapt the dedicated per‑resource implementation of:

[`IndexFilterTransformer`](https://github.com/camunda/camunda/blob/4d6edc8a85d03845759def87dd41b9f07181f6af/search/search-client-query-transformer/src/main/java/io/camunda/search/clients/transformers/filter/IndexFilterTransformer.java)

Specifically, adjust the `toAuthorizationCheckSearchQuery` method to:

- inspect `Authorization#resourceType`,
- derive the corresponding `resourceIds`, and build an appropriate query for expected resource type.

#### 2. RDBMS implementation

##### 2.1. DbQuery & DbReader

In the domain layer, extend the dedicated [`***DbQuery`](https://github.com/camunda/camunda/tree/main/db/rdbms/src/main/java/io/camunda/db/rdbms/read/domain) and [`***DbReader`](https://github.com/camunda/camunda/tree/main/db/rdbms/src/main/java/io/camunda/db/rdbms/read/service) so they:
- propagate per‑auth-resource-type authorized IDs via `ResourceAccessChecks.getAuthorizedResourceIdsByType(…)`,
- expose them as something like `authorizedResourceIdsByType` on the query object, typically replacing `authorizedResourceIds`,
  that will be passed to the MyBatis mapper.

##### 2.2. Mapper XML

Adjust the corresponding mapper XML under [`db/rdbms/src/main/resources/mapper`](https://github.com/camunda/camunda/tree/main/db/rdbms/src/main/resources/mapper)  
to use `authorizedResourceIdsByType` and express **disjunctive** SQL filters.

Previously, for a single authorization resource type (e.g `PROCESS_DEFINITION`), an authorization snippet might look like:

<details>
  <summary>Single <code>PROCESS_DEFINITION</code> resource type example</summary>

```xml
<!-- authorization filters -->
<if test="authorizedResourceIds != null and !authorizedResourceIds.isEmpty()">
  AND ut.PROCESS_DEFINITION_ID IN
  <foreach collection="authorizedResourceIds"
           item="resourceId" open="(" separator="," close=")">
    #{resourceId}
  </foreach>
</if>
```
</details>

To support both `PROCESS_DEFINITION` and `USER_TASK` with disjunctive semantics, you would migrate to:

<details>
  <summary>Multiple resource types (<code>PROCESS_DEFINITION</code> and <code>USER_TASK</code>)</summary>

```xml
<!-- authorization filters -->
<if test="authorizedResourceIdsByType != null and !authorizedResourceIdsByType.isEmpty()">
  AND
  <trim prefix="(" suffix=")" prefixOverrides="OR">
    <if test="authorizedResourceIdsByType['PROCESS_DEFINITION'] != null
              and !authorizedResourceIdsByType['PROCESS_DEFINITION'].isEmpty()">
      OR ut.PROCESS_DEFINITION_ID IN
      <foreach collection="authorizedResourceIdsByType['PROCESS_DEFINITION']"
               item="resourceId" open="(" separator="," close=")">
        #{resourceId}
      </foreach>
    </if>

    <if test="authorizedResourceIdsByType['USER_TASK'] != null
              and !authorizedResourceIdsByType['USER_TASK'].isEmpty()">
      OR ut.USER_TASK_KEY IN
      <foreach collection="authorizedResourceIdsByType['USER_TASK']"
               item="resourceId" open="("  separator="," close=")">
        #{resourceId}
      </foreach>
    </if>
  </trim>
</if>
```
</details>

Adopting this pattern for a given entity allows its RDBMS queries to correctly enforce composite (ANY) checks that span multiple `AuthorizationResourceType`s.

### Out of scope for this PR
This PR lays the foundation and wires it for user task queries, but does not change effective authorization behavior yet. Specifically:

- **User task checks are still enforced only via `PROCESS_DEFINITION[READ_USER_TASK]`.**  
  The `USER_TASK[READ]` part of the composite authorization for user tasks is **not yet enforced** and will be enabled in follow‑up PRs. Until then, checks remain identical to the existing behavior, based solely on `PROCESS_DEFINITION[READ_USER_TASK]`.

- **Only the user task path is prepared for multi‑type composites.**  
  Other `***FilterTransformer` implementations and `***DbReader`s are not adjusted in this PR.
  Additional resources will require their own adjustments (per the adoption guidelines above) before they can take advantage of composite checks involving different `AuthorizationResourceType`s.

## Related issues

closes #40126

ce-dmelnych committed:
feat: introduce authorization condition abstraction for composite checks

Add the `AuthorizationCondition` abstraction with `SingleAuthorizationCondition`
and `AnyOfAuthorizationCondition` to express composite (disjunctive) authorization.
All existing single-authorization flows will use `SingleAuthorizationCondition`
as a transparent wrapper, keeping current behavior unchanged.

This abstraction lays the foundation for supporting OR-based authorization evaluation
in the Search layer and enables future extensions such as ALL-OF or NOT conditions.

ce-dmelnych committed:
refactor: decouple AuthorizationChecker from SecurityContext

Refactor `AuthorizationChecker` to accept `CamundaAuthentication` and `Authorization<?>` directly
instead of `SecurityContext` in `retrieveAuthorizedAuthorizationScopes` and `isAuthorized` methods.

This prepares `SecurityContext` to carry `AuthorizationCondition` in follow-up commits.
Decoupling keeps `AuthorizationChecker` focused on single-authorization checks,
allowing future composite (ANY/OR) evaluation logic to be handled at a higher layer
without loading this class with authorization condition logic.

ce-dmelnych committed:
test: add doGet tests for AbstractResourceAccessController

Add unit tests for the `doGet` method in `AbstractResourceAccessController` covering happy path and edge cases:
- Resource returned when access is allowed
- `ResourceAccessDeniedException` thrown when resource access is denied
- `TenantAccessDeniedException` thrown when tenant access is denied

These tests establish baseline behavior before introducing composite authorization checks
(AnyOfAuthorizationCondition) support in follow-up commits.

ce-dmelnych committed:
fix: adjust TenantAccessDeniedException to include FORBIDDEN reason

Previously, TenantAccessDeniedException was created with the default
`UNKNOWN` reason. This change updates its construction to explicitly
use `Reason.FORBIDDEN`, aligning it with `ResourceAccessDeniedException`.

ce-dmelnych committed:
refactor: adjust SecurityContext to hold AuthorizationCondition

Adjust `SecurityContext` and `SecurityContextProvider` to hold/work with `AuthorizationCondition`
instead of only single authorization.

Adjust `SecurityContextProviderTest` to verify both `single` and `anyOf` conditions creation.

ce-dmelnych committed:
feat: enhance ResourceAccessDeniedException for multiple authorizations

Adjust `ResourceAccessDeniedException` to accept and represent multiple `Authorization` instances,
generating appropriate messages and `MissingAuthorization` entries.

This change will enable proper error reporting when evaluating composite (AnyOf) authorization conditions,
where none of the child authorizations are satisfied.

ce-dmelnych committed:
feat: add AnyOfAuthorizationCondition support for doGet operation

Enhance `AbstractResourceAccessController` to handle disjunctive (ANY) authorization checks
for `doGet` operations.

When processing `AnyOfAuthorizationCondition`, the controller iterates through all child
authorizations and grants access if any authorization is satisfied.
If none of the authorizations grant access, `ResourceAccessDeniedException` is thrown
with all failed authorizations.

Add comprehensive test coverage in `DocumentResourceAccessControllerTest` for
`AnyOfAuthorizationCondition` scenarios, including happy path (access granted by first/last authorization),
short-circuit evaluation (following authorization not checked), and denial case,
when all authorizations rejected with detailed error metadata.

Support for `doSearch` operations will be added in follow-up commits.

ce-dmelnych committed:
test: adjust test names in DocumentResourceAccessControllerTest

Refine `DocumentResourceAccessControllerTest` test names to explicitly reference which method
they are testing.

ce-dmelnych committed:
refactor: simplify creation of AuthorizationCheck

Remove unnecessary Optional-based logic from `createAuthorizationCheck(...)`
in favor of a more direct implementation.

Since all production code creates `ResourceAccess` with a valid
`Authorization` (nullable authorizations only appeared in a few tests),
the previous nullability checks were redundant.

To prevent such issues in the future, `ResourceAccess` now enforces that
its `Authorization` is never null via a constructor-level check.

ce-dmelnych committed:
feat: enhance AuthorizationCheck to hold AuthorizationCondition

Update `AuthorizationCheck` to hold `AuthorizationCondition` instead of a
single `Authorization`, preparing the codebase for upcoming support of
composite (AnyOf) authorization checks.

To keep the system compiling and preserve current behavior, callers
temporarily extract only the first authorization from the condition.
`IndexFilterTransformer` and `ResourceAccessChecks` will be updated to fully
support multiple authorizations in subsequent commits.

ce-dmelnych committed:
feat: add AnyOfAuthorizationCondition support for doSearch operation

Enhance `AbstractResourceAccessController` to handle `AnyOfAuthorizationCondition` during `doSearch`
by iterating over child authorizations, short-circuiting on wildcard `ResourceAccess` to disable checks,
and otherwise aggregating all resolved authorizations into a single `AuthorizationCheck`.

ce-dmelnych committed:
feat: add AnyOfAuthorizationCondition support to IndexFilterTransformer

Extend IndexFilterTransformer to evaluate composite (disjunctive)
authorization conditions. The transformer now distinguishes between
`SingleAuthorizationCondition` and `AnyOfAuthorizationCondition`, applying
OR-based query construction when multiple authorization branches are
present.

This enables correct ES/OS-side handling of composite authorization checks.

ce-dmelnych committed:
feat: extend UserTaskFilterTransformer to support USER_TASK checks

Update `UserTaskFilterTransformer#toAuthorizationCheckSearchQuery` to handle
additional authorization resource types when searching for user tasks.
In addition to the existing `PROCESS_DEFINITION` case, the transformer
now maps `USER_TASK` authorizations to queries on the `KEY` field.

USER_TASK authorization `resourceIds` are parsed to long since the `key`
field in ES/OS `task` template mappings is of type long.

This enables the Search layer to evaluate user-task-level permissions and
lays the groundwork for enforcing `PROCESS_DEFINITION[READ_USER_TASK]` and
`USER_TASK[READ]` checks wrapped by `AnyOfAuthorizationCondition` in
following PRs.

ce-dmelnych committed:
feat: add composite-aware resourceId resolution in ResourceAccessChecks

Introduce `getAuthorizedResourceIdsByType()`, which returns a mapping from
resource type to the merged list of authorized resource IDs for same type across all
authorization branches. This method guarantees that only resource types with at least
one resource ID appear in the result and will be used by composite-aware readers.

Update `AbstractEntityReader` to rely on `getAuthorizedResourceIdsByType()`,
which is safe for both single and composite authorization conditions.

Add JavaDoc to `getAuthorizedResourceIds()`, clarifying that it returns resource
IDs only for the *first* authorization branch. This method remains available for
`***DbReader` classes that still operate under the single-authorization
assumption and will be replaced by `getAuthorizedResourceIdsByType()` once
readers need to support composite authorization checks.

Add test coverage for both methods to validate their semantics and ensure
backwards compatibility while preparing the codebase for composite authorization
support.

ce-dmelnych committed:
feat: introduce dedicated authorization fields in UserTaskDbQuery

Extend `UserTaskDbQuery` to store authorization data using two explicit,
type-safe fields:
- `authorizedProcessDefinitionIds : List<String>`
- `authorizedUserTaskKeys : List<Long>`

These fields replace the former single-authorization `authorizedResourceIds`
mechanism and provide a clearer structure for supporting composite (ANY-OF)
authorization conditions during user-task searches.

Update `UserTaskDbReader` to populate these lists based on the resolved
authorization condition. User-task resource IDs are parsed to `Long` as
`ut.USER_TASK_KEY` has a BIGINT typevia, this ensures cross-database compatibility
(I found that `PostgreSQL` does not implicitly convert string literals to BIGINT and throws error),
so normalizing values at the Java layer avoids SQL casting issues and
prevents invalid (non-numeric) IDs from reaching the query layer.

A follow-up commit will update `UserTaskMapper.xml` to build authorization
predicates using these new per-type fields.

ce-dmelnych committed:
feat: update UserTaskMapper to apply per-type authorization filters

Adjust UserTaskMapper.xml `searchFilter` mapping to use the new dedicated
authorization fields introduced in `UserTaskDbQuery`:
- `authorizedProcessDefinitionIds : List<String>`
- `authorizedUserTaskKeys : List<Long>`

The previous mapping based on `authorizedResourceIds` has been replaced with a
per-resourceId-property authorization filter that evaluates `processDefinitionIds` and
`userTaskKeys` branches separately and combines them inside a single parenthesized `OR` expression.
This allows SQL translation of composite (ANY-OF) authorization conditions.

Examples of generated SQL for different `authorized<resourceId>` inputs:
# 1. PROCESS_DEFINITION-only authorization
Input:
> { "authorizedProcessDefinitionIds": ['procA', 'procB']}

SQL:
> AND ( ut.PROCESS_DEFINITION_ID IN ('procA','procB'))

# 2. USER_TASK-only authorization
Input:
> {"authorizedUserTaskKeys": [123, 456]}

SQL:
> AND (ut.USER_TASK_KEY IN (123,456))

# 3. Both PROCESS_DEFINITION and USER_TASK
Input:
> {
>    "authorizedProcessDefinitionIds": ['procX'],
>    "authorizedUserTaskKeys": [789]
> }

SQL:
> AND (ut.PROCESS_DEFINITION_ID IN ('procX') OR ut.USER_TASK_KEY IN (789))

# 4. No authorization provided -> no filter emitted
Input:
> {
>    "authorizedProcessDefinitionIds": [],
>    "authorizedUserTaskKeys": []
> }

SQL:
> (no authorization predicate added)

ce-dmelnych committed:
test: require AuthorizationResourceType in search tests setup

Update the `resourceAccessChecksFromResourceIds` test util methods to require an
explicit `AuthorizationResourceType` when constructing `ResourceAccessChecks`.
Previously, tests created `ResourceAccessChecks` only from resource IDs, without
specifying the resource type—something that never occurs in real usage.

All affected tests are updated to provide the correct resource type, bringing the
test setup in line with production behavior, where every `Authorization` instance
always carries an associated `AuthorizationResourceType`.

For `ClusterVariableIT`, the tests use `AuthorizationResourceType.UNSPECIFIED`
because cluster variables do not yet have a dedicated authorization resource type.

This ensures more realistic test conditions and aligns the codebase with recent
changes to support composite authorisation handling.

ce-dmelnych committed:
feat: verify composite authorization checks for user task search

Add test coverage for composite authorization checks in UserTaskIT.
The new tests verify that the user task search correctly handles disjunctive (ANY)
authorization conditions with multiple resource types

ce-dmelnych commented:
@copilot update "Pull request overview" comment (https://github.com/camunda/camunda/pull/41381#pullrequestreview-3499199499) based on all commits in this PR.

romansmirnov reviewed (commented):
Overall, the changes look good. I like the approach. I left a few comments.

ce-dmelnych committed:
refactor: deduplicate authorization handling and improve logging

Reuse a shared single `Authorization<?>` helper so both `Single` and `AnyOf` authorization conditions consistently
short-circuit when resource identifiers are missing, which avoids building redundant `matchNone` fragments.
Skipping the `matchNone` results during `AnyOf` authrizations handling keeps resulting queries minimal.

Emitting a warning when `UserTaskFilterTransformer` encounters an unsupported authorization resource type
clarifies why a request falls back to `matchNone` and highlights that only
`PROCESS_DEFINITION` and `USER_TASK` are supported.

ce-dmelnych committed:
feat: add `hasAnyResourceId()` to ResourceAccessChecks and use it

Introduce `hasAnyResourceId()` in `ResourceAccessChecks` to provide an efficient
boolean check for whether any authorization branch contains at least one
resource ID.

Update `***DbReader#shouldReturnEmptyResult` to rely on this new method instead
of `getAuthorizedResourceIds()`, which was tied to single-authorization
semantics and triggered unnecessary data extraction.

ce-dmelnych committed:
feat: migrate Readers to composite-aware getAuthorizedResourceIdsByType

Update all `***DbReader` classes to use `ResourceAccessChecks#getAuthorizedResourceIdsByType()`
instead of `getAuthorizedResourceIds()`.

The removed method only handled single-branch authorizations, while `getAuthorizedResourceIdsByType()`
supports composite conditions and separates IDs by `AuthorizationResourceType`.

`***DbReader`s now extract resource IDs per type (e.g., TENANT, PROCESS_DEFINITION, USER_TASK)
from the map returned by `getAuthorizedResourceIdsByType()`, preparing the readers for
composite authorization support going forward.

romansmirnov reviewed (approved):
🚀 Thanks for applying the review feedback. LGTM.

ce-dmelnych committed:
fix: update AuditLogDbReader to use composite-aware API

`AuditLogDbReader` was referencing the removed `ResourceAccessChecks#getAuthorizedResourceIds()` method,
causing a compilation failure.

`AuditLogDbReader` currently do not support resource-ID–based authorization constraints,
so this reader now resolves authorized resource IDs via `getAuthorizedResourceIdsByType()`
and falls back to an empty list under `AuthorizationResourceType.UNSPECIFIED`
until a dedicated resource type is introduced (tracked in #41120).

romansmirnov reviewed (approved):
LGTM 😄

npepinpe reviewed (commented):
Note I only looked at the security module, and a bit how the new things are used. I didn't do an exhaustive review of how it's implemented for the search and RDBMS layers.

Generally it looks fine, but the solution likely doesn't extend very well if we ever want different kinds of conditions. It seems like you originally wanted to delegate to the implementation of `AuthorizationCondition` to decide how this is defined, but ultimately opted to just have this be a carrier for a list of authorizations? What's the point of the interface in this case, and why not simply pass a list? Am I missing something? How would we extend this to, say, conjunctive conditions?
